### ISSUE TITLE
Edge empty area around sampling and padding effect

### ISSUE THREAD

**Issue by SnipG** [Scirra/Construct-3-bugs#4431]
## Problem description
Case I don't know where to post. Here or as feature suggestion.

If user adds padding to sprite from animation editor, then there is constant wasted space around it. And if there is no padding, edges are not sampled. 

Look at image below. Should'nt sprite edge come after sampling. Currently there is always like 1 pixel border around sprite from which half is sampled and half is always empty. Depending on case and scale, the empty area can be large and lower performance a lot.

I know this might be how sampling works and always samples half of edge pixels, which leaves behaind empty wasted area.
![edgeCutOffff](https://user-images.githubusercontent.com/28864889/103041968-a4ec1d80-4580-11eb-9488-5b1d56170865.jpg)

And if it is, pixel half is mixed with nearest pixel half, then just export or cut image padding 0.5px instead. Which removes empty area

Side by side with nearest sampling.

![AnotherCase3](https://user-images.githubusercontent.com/28864889/103043766-bafcdc80-4586-11eb-9789-70b51cf5e316.jpg)

## Attach a .c3p
[edge_sampling.zip](https://github.com/Scirra/Construct-3-bugs/files/5737387/edge_sampling.zip)

[red_dot_edge.zip](https://github.com/Scirra/Construct-3-bugs/files/5739953/red_dot_edge.zip)

## Steps to reproduce

1.
2.
3.

## Observed result
There always seems  to be some empty area around sprite.  Around half a pixel, but on scaled images this is large.

## Expected result
If it's bug, could it be removed. 0.5 pixel cut off would help 50% nearest sampling and 100% bilinear and trilinear

## More details

**Affected browsers/platforms:** 

**First affected release:** 

## System details

<details><summary>View details</summary>

Platform information
Browser: Chrome
Browser version: 87.0.4280.88
Browser engine: Chromium
Browser architecture: 64-bit
Context: browser
Operating system: Windows
Operating system version: 10
Operating system architecture: 64-bit
Device type: desktop
Device pixel ratio: 1
Logical CPU cores: 12
Approx. device memory: 8 GB
User agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.88 Safari/537.36
C3 release: r226 (beta)
Language setting: en-US

Local storage
Storage quota (approx): 120 gb
Storage usage (approx): 1.4 gb (1.1%)
Persistant storage: Yes

Browser support notes
This list contains missing features that are not required, but could improve performance or user experience if supported.

Nothing is missing. Everything is OK!
WebGL information
Version string: WebGL 2.0 (OpenGL ES 3.0 Chromium)
Numeric version: 2
Supports NPOT textures: yes
Supports GPU profiling: yes
Supports highp precision: yes
Vendor: Google Inc.
Renderer: ANGLE (NVIDIA GeForce GTX 660 Ti Direct3D11 vs_5_0 ps_5_0)
Major performance caveat: no
Maximum texture size: 16384
Point size range: 1 to 1024
Extensions:

EXT_color_buffer_float
EXT_color_buffer_half_float
EXT_disjoint_timer_query_webgl2
EXT_float_blend
EXT_texture_compression_bptc
EXT_texture_compression_rgtc
EXT_texture_filter_anisotropic
EXT_texture_norm16
KHR_parallel_shader_compile
OES_texture_float_linear
WEBGL_compressed_texture_s3tc
WEBGL_compressed_texture_s3tc_srgb
WEBGL_debug_renderer_info
WEBGL_debug_shaders
WEBGL_lose_context
WEBGL_multi_draw
OVR_multiview2
Audio information
System sample rate: 48000 Hz
Output channels: 2
Output interpretation: speakers
Supported decode formats:

WebM Opus (audio/webm; codecs=opus)
Ogg Opus (audio/ogg; codecs=opus)
WebM Vorbis (audio/webm; codecs=vorbis)
Ogg Vorbis (audio/ogg; codecs=vorbis)
MPEG-4 AAC (audio/mp4; codecs=mp4a.40.5)
MP3 (audio/mpeg)
FLAC (audio/flac)
PCM WAV (audio/wav; codecs=1)
Supported encode formats:

WebM Opus (audio/webm; codecs=opus)
Video information
Supported decode formats:

WebM AV1 (video/webm; codecs=av01.0.00M.08)
MP4 AV1 (video/mp4; codecs=av01.0.00M.08)
WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)
Ogg Theora (video/ogg; codecs=theora)
H.264 (video/mp4; codecs=avc1.42E01E)
Supported encode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)

</details>

**Comment by AshleyScirra** [comment]
This is by design. The image editor crop tool leaves a 1px transparent border in order to improve the quality of edges when rotating and scaling sprites, and be consistent with edge quality with interior transparency. Since nothing is drawn outside the object box, there needs to be a 1px border to add a 0.5px gradient in linear sampling mode. This doesn't apply to nearest sampling, but the overhead is negligible.

**Comment by SnipG** [comment]
Would it be possible to reduce overhead by a feature, if it's by the design? If so, I would file one.
Overhead can be in some case as big as 30%. When rotate is needed and when scaled or zoomed, 0.5px gets huge.

Case that something scales or crops sprite internally and not with fragment shader which cuts performance ~2x. Or webgl own operations equal fragment shader performance drops?

**Comment by AshleyScirra** [comment]
The overhead is negligible. Consider a 100x100 image that has 10000 pixels. Padding it to 102x102 makes it 10404 pixels. That's just a 4% increase. The proportion of the overhead will be even less for larger images.

**Comment by SnipG** [comment]
I made the issue report, just to look if there is other ways around and was just curious if possible some internal change is viable before feature suggestion. I used 15x15 to 25x25 objects. With them I used shader to scale 25x25 up to 100+, but keep image clear with edge sampling fragment shader.
But here sampling adds 2x performance impact. So I went to ~50x50 instead and without shader, they scale well to, better then 25x25. 

Just looking game examples on c3 page, for game objects: stars, stones, enemies. Little bullets. Average size is like 25x25 - 60x60.
Lets take 40x40 to 42x42 its 10%.
Glokar example has  bullets 15x15, they dont use border, but if one is to add it, they give 28% increase. 

If user has small sprites which are dominating, it affects. And ofc for big sprites it's super small, but for little ones you can clearly see the transparent area taking up huge area. 

Ofc it's pointless when:
  *you use nearest sampling you don't have to use padding at all
  *Also when you don't need to rotate objects. No sampling needed

Case I had is more like some edge case, you use small images mostly, need sampling, rotate and scale and use a lot of them.

**Comment by AshleyScirra** [comment]
It should have negligible performance impact. I cannot imagine how it could have a "2x performance impact". Small sprites have a tiny impact on fillrate so even if there is a proportionally larger impact to fillrate, then it is still basically negligible in absolute terms because the sprite is so small. You should demonstrate actual performance benchmarks to prove your point, since I can't see any way this would make a meaningful difference to performance.

You can also hold control and click crop to not add the 1px transparent border. This will adjust the blending at the object edges - notably making a jaggy, aliased edge even with linear sampling, which is what this feature is intended to avoid - but it should have no significant impact to performance.

**Comment by SnipG** [comment]
I used workaround for this case, just using bigger image where transparent % is smaller.

I don't have test, but to summarize current issue. 
I looked up the case when I wondered if 1 px border should/could be different:
7x7,  7x15
![edge2](https://user-images.githubusercontent.com/28864889/103577312-38a4de80-4edd-11eb-931d-9dacf9bed289.jpg)

With pixel size and shader effect. I can scale 7x15 for example to 718 x 1536.428571.  Without using such large texture and has sampling.

But overall:
Would vertex shader able remove transparent area with almost no cost?
Would some webgl command remove transparent area?

Or they both same costly as fragment shader and one should use it? Currently fragment shader could impact more then transparent area itself does.

If one of two is feasable i could just support or file feature suggestion.
